File: src/common/py/utils/config/setting_id.py
import dataclasses
import typing


@dataclasses.dataclass(frozen=True)
class SettingID:
    category: str | None
    name: str
    
    def split(self) -> typing.List[str]:
        return str(self).split(".")
    
    def env_name(self, prefix: str) -> str:
        return f"{prefix}_{str(self).replace('.', '_')}".upper()
    
    def __str__(self) -> str:
        return f"{self.category}.{self.name}" if self.category is not None else self.name
    
    def __repr__(self) -> str:
        return str(self)

File: src/common/web/utils/config/SettingID.ts
/**
 * A setting identifier.
 *
 * Settings are specified by a category they belong to, as well as their actual name.
 *
 * Categories support sub-categories by separating them using dots (.);
 * when represented as a string, all component tokens are separated by dots.
 */
export class SettingID {
    public constructor(readonly category: string, readonly name: string) {
    }

    /**
     * Splits the entire identifier into single string tokens.
     *
     * @returns - The tokens list.
     */
    public split(): string[] {
        return this.toString().split(".");
    }

    /**
     * Generates an environment variable name for this identifier.
     *
     * A setting identifier is translated to its corresponding environment variable name by replacing all dots (.) with underscores (_),
     * prepending a ``prefix``, as well as making everything uppercase.
     *
     * @param prefix - prefix: The prefix to prepend.
     *
     * @returns - The corresponding environment variable name.
     */
    public envName(prefix: string): string {
        return `${prefix}_${this.toString().replace(".", "_")}`.toUpperCase();
    }

    /**
     * Converts the setting ID to a string.
     *
     * @returns - The string representation of this setting ID.
     */
    public toString(): string {
        return this.category ? `${this.category}.${this.name}` : this.name;
    }
}
